### PR TITLE
fix(cli): cdk watch for Lambdas with Advanced Logging Controls do not stream logs to the terminal

### LINKED ISSUE #29448: (cdk cli): cdk watch with custom log groups do not stream logs to the terminal
### Describe the bug

After the release of Advanced Logging Controls for Lambda (see https://aws.amazon.com/blogs/compute/introducing-advanced-logging-controls-for-aws-lambda-functions/) I've decided to move all the logs of my multi-stack deployment to a single unified Log Group per deployed tenant.

It goes likes this:

- The main stack creates the LogGroup like `/aws/lambda/{tenant}`;
- The secondary stacks refers to the LogGroup using ` LogGroup.fromLogGroupArn(...);`

Then I discovered that running `cdk watch` on the main stack I can see the cloudwatch logs on my terminal while on the secondary stacks it does not;

I found at that the cloudwatch log group resolver for the logs basically just assumes the log group is `/aws/lambda/{physicalId}`

https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk/lib/api/logs/find-cloudwatch-logs.ts#L114

### Expected Behavior

I expect the logs to the displayed on terminal for lambda functions using pre-existing CloudWatch Log Groups not created in the stack (referred by LogGroup.fromLogGroupArn)

### Current Behavior

I don't see the logs.

### Reproduction Steps

Create a stack that creates a log group;

```
new LogGroup(this, 'tenant-log-group', {
        logGroupName: `/aws/lambda/xxx`
});
```

On a secondary stack create a Lambda Function that uses the `logGroup` referred by `LogGroup.fromLogGroupArn` instead of creating the log group in the stack itself;

### Possible Solution

Check the `LoggingConfig.LogGroupName` property of the lambda to resolve the CloudWatch Log Group of the Lambda

### Additional Information/Context

_No response_

### CDK CLI Version

2.131.0 (build 92b912d)

### Framework Version

_No response_

### Node.js Version

20

### OS

MacOS

### Language

TypeScript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by onhate**
### Issue

Closes #29448

### Reason for this change

After the release of Advanced Logging Controls for Lambda (see https://aws.amazon.com/blogs/compute/introducing-advanced-logging-controls-for-aws-lambda-functions/) I've decided to move all the logs of my multi-stack deployment to a single unified Log Group per deployed tenant.

It goes likes this:

The main stack creates the LogGroup like `/aws/lambda/{tenant}`;
The secondary stacks refers to the LogGroup using  `LogGroup.fromLogGroupArn(...)`;
Then I discovered that running cdk watch on the main stack I can see the cloudwatch logs on my terminal while on the secondary stacks it does not;

I found at that the cloudwatch log group resolver for the logs basically just assumes the log group is `/aws/lambda/{physicalId}`

https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk/lib/api/logs/find-cloudwatch-logs.ts#L114

### Description of changes

Use the Template JSON to resolve the `LoggingConfig.LogGroupName` of the Lambda Function;

### Description of how you validated changes

I've replace the node_modules/aws-cdk of my project with the one built from this PR and I was able to see the logs of my lambda that has custom LogConfig, as you can see it is grabbing the logs of `/aws/lambda/dev` cloudwatch logs

2024-03-13T10-46-48

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

### CODE REVIEW

**Comment by onhate** [comment]
Exemption Request

as this change is only on the `cdk watch` logic I could not find a way to write a integration test for it.

File: packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts [hunk 1]
```diff
@@ -314,6 +314,10 @@ export class EvaluateCloudFormationTemplate {
     return cfnExpression;
   }
 
+  public getResourceProperty(logicalId: string, propertyName: string) {
```
Comment (msambol) [inline packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts:317]: can you add a return type ?

File: packages/aws-cdk/lib/api/logs/find-cloudwatch-logs.ts [hunk 2]
```diff
@@ -84,38 +74,68 @@ function isReferencedFromIgnoredResource(
...
+    return resource.PhysicalResourceId?.toString();
+  },
+
+  // Resource types that will create a CloudWatch log group with a specific name if one is not provided.
+  // The keys are CFN resource types, and the values are the name of the physical name property of that resource
+  // and the service name that is used in the automatically created CloudWatch log group.
+  'AWS::Lambda::Function': (resource, evaluateCfnTemplate) => {
+    const loggingConfig = evaluateCfnTemplate.getResourceProperty(resource.LogicalResourceId, 'LoggingConfig');
+    if (loggingConfig?.LogGroup) {
+      // if LoGroup is a string then use it as the LogGroupName as it is referred by LogGroup.fromLogGroupArn in CDK
```
Comment (msambol) [inline packages/aws-cdk/lib/api/logs/find-cloudwatch-logs.ts:102]: s/LoGroup/LogGroup

**Comment by onhate** [comment]
@msambol done :)

File: packages/aws-cdk/lib/api/logs/find-cloudwatch-logs.ts [hunk 3]
```diff
@@ -84,38 +74,68 @@ function isReferencedFromIgnoredResource(
...
+    if (isReferencedFromIgnoredResource(resource, evaluateCfnTemplate)) {
+      return undefined;
+    }
+    return resource.PhysicalResourceId?.toString();
+  },
+
+  // Resource types that will create a CloudWatch log group with a specific name if one is not provided.
+  // The keys are CFN resource types, and the values are the name of the physical name property of that resource
+  // and the service name that is used in the automatically created CloudWatch log group.
+  'AWS::Lambda::Function': (resource, evaluateCfnTemplate) => {
```
Comment (scanlonp) [inline packages/aws-cdk/lib/api/logs/find-cloudwatch-logs.ts:99]: Lambda is not the only service that creates custom `logGroupName`s. We should note that similar functionality may be needed for several other services (rds, neptune, docdb, etc). However, Lambda Function was the only resource in `RESOURCE_TYPES_WITH_IMPLICIT_LOGS`, so we definitely do not need to add that functionality.

**Review by scanlonp — CHANGES_REQUESTED** [review]
Thanks for this feature! It looks very good, and the in-line documentation is fantastic. I just had a couple nits and clarifications.

The only corner case I was thinking about would be handling tokens, but since we are referencing log groups from another template, I think those should be resolved.

I think your validation is sufficient to not need the integ test.

**Comment by scanlonp** [comment]
@Mergifyio update

**Comment by scanlonp** [comment]
I will submit this to the test pipeline as well.

File: packages/aws-cdk/lib/api/logs/find-cloudwatch-logs.ts (hunk 3 above)
Comment (onhate) [reply packages/aws-cdk/lib/api/logs/find-cloudwatch-logs.ts:99]: I though about that while making this change but as the only reference I had was the Lambda one I thought in not adding more stuff than the things that already exists, but I can definitely work on that extras on upcoming PRs now that I got familiarity with the codebase.

File: packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts [hunk 4]
```diff
@@ -314,6 +314,10 @@ export class EvaluateCloudFormationTemplate {
     return cfnExpression;
   }
 
+  public getResourceProperty<T = any>(logicalId: string, propertyName: string): T {
```
[the remark this answers is not in the document]
Comment (onhate) [reply packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts:317]: I thought in allowing the user pass a cast type if the user already know the type but also have any for all other, but can remove it for sure.

File: packages/aws-cdk/lib/api/logs/find-cloudwatch-logs.ts [hunk 5]
```diff
@@ -84,38 +74,68 @@ function isReferencedFromIgnoredResource(
...
   const resourcesReferencingLogGroup = evaluateCfnTemplate.findReferencesTo(logGroupResource.LogicalResourceId);
-  for (const reference of resourcesReferencingLogGroup) {
-    if (IGNORE_LOGS_RESOURCE_TYPES.includes(reference.Type)) {
-      foundReference = true;
-    }
-  }
-  return foundReference;
+  return resourcesReferencingLogGroup.some(reference => {
+    return IGNORE_LOGS_RESOURCE_TYPES.includes(reference.Type);
+  });
```
[the remark this answers is not in the document]
Comment (onhate) [reply packages/aws-cdk/lib/api/logs/find-cloudwatch-logs.ts:80]: correct, it was looping the resourcesReferencingLogGroup and returning false if some Type is in the IGNORE_LOGS_RESOURCE_TYPES.

Which summarizes as a `.some.

```ts
let foundReference = false;
  const resourcesReferencingLogGroup = evaluateCfnTemplate.findReferencesTo(logGroupResource.LogicalResourceId);
  for (const reference of resourcesReferencingLogGroup) {
    if (IGNORE_LOGS_RESOURCE_TYPES.includes(reference.Type)) {
      foundReference = true;
    }
  }
  return foundReference;
 ```

**Comment by scanlonp** [comment]
Hey @onhate, thanks for the quick update. I ran this through the pipeline, but it is failing due to another change we have made. Once I fix the pipeline, I'll throw this in there again, and if it passes, approve it!

**Comment by scanlonp** [comment]
Also, can you tweak your PR title so that it describes the bug rather than the fix? Along the lines of "Referenced lambda streams not used for cdk watch", but more eloquent.

**Comment by onhate** [comment]
@scanlonp do you have an update on this matter? thanks.

**Comment by scanlonp** [comment]
@onhate, I should be good to submit it later today. I will keep you updated on the status of the run.

**Comment by scanlonp** [comment]
@Mergifyio update

**Review by scanlonp — APPROVED** [review]
Passing the test pipeline. Thanks for your patience!

**Pull request merged**